### PR TITLE
CfOverride

### LINKED ISSUE #9104: Removing `cl_overrides`
Given that we have `cf_flags` now, we could utilize it to replace `cl_overrides` without additional memory cost. For the most part, this comes down to replacing some `List.memq` with a flag check. However, there are some more involved operations on `cl_overrides` in the dark realms of gencommon which we would have to check.

For macro compatibility, we would have to iterate `cl_ordered_fields` and check for the override flag. This is more expensive, but I doubt that it could have a significant impact even for classes with hundreds of fields. Still, we should check this to make sure.

What do you think?

### PR DESCRIPTION
**Pull request by Simn**
closes #9104
see #9183

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
This is going to be a lot of fun with gencommon

**Comment by Simn** [comment]
asdfg...

File: src/context/display/statistics.ml [hunk 1]
```diff
@@ -52,22 +52,20 @@ let collect_statistics ctx pfilter with_expressions =
...
-						()
-					end;
-					loop c
-				| _ ->
+	let check_override c cf =
+		let rec loop c = match c.cl_super with
+			| Some (c,_) ->
+				begin try
+					let cf' = PMap.find cf.cf_name c.cl_fields in
+					add_relation cf'.cf_name_pos (Overridden,cf.cf_name_pos)
```
Comment (RealyUniqueName) [inline src/context/display/statistics.ml:60]: Missing a check for `CfOverride` here

File: src/context/display/statistics.ml [hunk 2]
```diff
@@ -242,7 +239,10 @@ let collect_statistics ctx pfilter with_expressions =
 				end
 			in
 			Option.may field c.cl_constructor;
-			List.iter field c.cl_ordered_fields;
+			List.iter (fun cf ->
+				check_override c cf;
```
Comment (RealyUniqueName) [inline src/context/display/statistics.ml]: Or here

Comment (Simn) [reply src/context/display/statistics.ml]: Yep here, good call!

File: src/generators/gencs.ml [hunk 3]
```diff
@@ -2475,7 +2475,7 @@ let generate con =
 			(try
 				if cl.cl_interface then raise Not_found;
 				let cf = PMap.find "finalize" cl.cl_fields in
-				(if List.exists (fun c -> c.cf_name = "finalize") cl.cl_overrides then raise Not_found);
+				(if List.exists (fun c -> c.cf_name = "finalize" && has_class_field_flag c CfOverride) cl.cl_ordered_fields then raise Not_found);
```
Comment (RealyUniqueName) [inline src/generators/gencs.ml]: No need for `List.exists (fun c -> c.cf_name = "finalize"` anymore as it's already "checked" by `PMap.find "finalize"` in the previous line.

File: src/generators/gencs.ml [hunk 4]
```diff
@@ -2455,7 +2455,7 @@ let generate con =
 			(try
 				if cl.cl_interface then raise Not_found;
 				let cf = PMap.find "toString" cl.cl_fields in
-				(if List.exists (fun c -> c.cf_name = "toString") cl.cl_overrides then raise Not_found);
+				(if List.exists (fun c -> c.cf_name = "toString" && has_class_field_flag c CfOverride) cl.cl_ordered_fields then raise Not_found);
```
Comment (RealyUniqueName) [inline src/generators/gencs.ml]: Same here (as in the next comment).

File: src/generators/gencs.ml (hunk 3 above)
Comment (Simn) [reply src/generators/gencs.ml]: I was worried about overloads, but I guess even in that case there's only one in `cl_ordered_fields`.

File: src/generators/genhl.ml [hunk 5]
```diff
@@ -4003,7 +4003,7 @@ let add_types ctx types =
 				| _ ->
 					false
 			in
-			if not ctx.is_macro then List.iter (fun f -> ignore(loop c.cl_super f)) c.cl_overrides;
+			if not ctx.is_macro then List.iter (fun f -> if has_class_field_flag f CfOverride then ignore(loop c.cl_super f)) c.cl_ordered_fields;
```
Comment (RealyUniqueName) [inline src/generators/genhl.ml:4006]: Shouldn't this also traverse `cl_ordered_statics`?

Comment (Simn) [reply src/generators/genhl.ml:4006]: No because overrides are never static.

File: src/generators/genjvm.ml [hunk 6]
```diff
@@ -2249,7 +2249,8 @@ class tclass_to_jvm gctx c = object(self)
 			) c_int.cl_implements
 		in
 		loop (fun t -> t) c;
-		begin match c.cl_overrides,c.cl_super with
+		let overrides = List.filter (fun cf -> has_class_field_flag cf CfOverride) c.cl_ordered_fields in
```
Comment (RealyUniqueName) [inline src/generators/genjvm.ml:2252]: `cl_ordered_statics`?

Comment (RealyUniqueName) [reply src/generators/genjvm.ml:2252]: There are quite a few other places, where `cl_ordered_statics` is probably required too.

Comment (RealyUniqueName) [reply src/generators/genjvm.ml:2252]: Also, what about `cl_constructor`? I think they also could be overloaded.

Comment (nadako) [reply src/generators/genjvm.ml:2252]: Huh? Isn't this about overrides, not _overloads_?

File: src/generators/genhl.ml (hunk 5 above)
Comment (RealyUniqueName) [reply src/generators/genhl.ml:4006]: Ah, I'm always confused with "override"/"overload". These words are too similar for me in english )

File: src/generators/genjvm.ml (hunk 6 above)
Comment (RealyUniqueName) [reply src/generators/genjvm.ml:2252]: Yeah, I'm just always confused by these words similarity )

File: src/generators/gencs.ml (hunk 3 above)
Comment (RealyUniqueName) [reply src/generators/gencs.ml]: Yes, AFAIR overloads are not pushed to `cl_ordered_fields`

Comment (Simn) [reply src/generators/gencs.ml]: From the typer yes, but I wasn't sure what dark arts gencommon might be performing.

Anyway, looks like this is fine.

**Comment by Simn** [comment]
I think this would pass CI now if CI wasn't failing for other reasons. I'll merge it if Azure agrees.

**Pull request merged**